src/tiled/changemapobject.cpp

```diff
@@ -99,35 +99,39 @@ ChangeMapObjectsTile::ChangeMapObjectsTile(MapDocument *mapDocument,
     , mMapObjects(mapObjects)
     , mTile(tile)
 {
+    mCell = Cell(mTile);
     for (MapObject *object : mMapObjects) {
-        Tile *tile = object->cell().tile();
-        mOldTiles.append(tile);
-        mSizePreserved.append(object->size() == tile->size());
+        Cell cell = object->cell();
+        mOldCells.append(cell);
+        mOldSizes.append(object->size());
+        Tile *tile = cell.tile();
+        mUpdateSize.append(tile == nullptr || object->size() == tile->size());
```

Comment:
If the previous tile was no longer valid, you don't know whether the object used to match the size or not. I think the decision about whether to update the size is arbitrary in that case, and I may have chosen to leave the size alone instead. That way you also don't need `mOldSizes`.

Please leave out explicit comparisons to `nullptr`, btw. You can just write `!tile`, or if you would change the condition, `tile && object->size() == tile->size()`.